core/connection.js

```diff
@@ -675,3 +697,103 @@ Blockly.Connection.prototype.toString = function() {
   }
   return msg + block.toDevString();
 };
+
+/**
+ * Returns the state of the shadowDom_ and shadowState_ properties, then
+ * temporarily sets those properties to null so no shadow respawns.
+ * @return {{shadowDom: ?Element, shadowState: ?Object}} The state of both the
+ *     shadowDom_ and shadowState_ properties.
+ */
+Blockly.Connection.prototype.stashShadowState_ = function() {
+  const shadowDom = this.getShadowDom(true);
+  const shadowState = this.getShadowState(true);
+  // Set to null so it doesn't respawn.
+  this.shadowDom_ = null;
+  this.shadowState_ = null;
+  return {shadowDom, shadowState};
+};
+
+/**
+ * Reapplies the stashed state of the shadowDom_ and shadowState_ properties.
+ * @param {{shadowDom: (?Element|undefined), shadowState:
+ *     (?Object|undefined)}=} param0 The state to reapply to the shadowDom_ and
+ *     shadowState_ properties.
+ */
+Blockly.Connection.prototype.applyShadowState_ =
+    function({shadowDom, shadowState}) {
+      this.shadowDom_ = shadowDom;
+      this.shadowState_ = shadowState;
+    };
+
+/**
+ * Sets the state of the shadow of this connection.
+ * @param {{shadowDom: (?Element|undefined), shadowState:
+ *     (?Object|undefined)}=} param0 The state to set the shadow of this
+ *     connection to.
+ */
+Blockly.Connection.prototype.setShadowStateInternal_ =
+    function({shadowDom = null, shadowState = null} = {}) {
+      // One or both of these should always be null.
+      // If neither is null, the shadowState will get priority.
+      this.shadowDom_ = shadowDom;
+      this.shadowState_ = shadowState;
+
+      var target = this.targetBlock();
+      if (!target) {
```

Comment:
Could you explain these 3 different cases? Not necessarily in code comments yet but just I don't understand why each case does what it does.

Reply:
I can give it my best shot haha

1) `!target` means that there is no connected block (no shadow and no real). In this case we want to try to respawn a shadow. If one is created, we serialize it.
2) `target.isShadow()` means there is a connected shadow block. In this case we want to dispose of that and try to respawn a new shadow. If one is created, we serialize it.
3) Otherwise we know we have a target block (b/c `!target` was false) and we know it's not a shadow, so we must have a connected real block. In this case we just want to create the shadow block (without trying to connect it) and serialize it.

The serializing is to make sure that our `shadowDom_` *and* `shadowState_` properties both have the correct value so that both serialization systems work properly.

Tell me if there's anything I didn't explain well. Cuz honestly all of this logic is a bit of a (hopefully necessary?) mess.

And if there's anything I can do to clean this up, definitely tell me that as well.